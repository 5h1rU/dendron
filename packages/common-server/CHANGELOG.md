# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.24](https://github.com/dendronhq/dendron/compare/v0.3.23...v0.3.24) (2020-07-20)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.21](https://github.com/dendronhq/dendron/compare/v0.3.20...v0.3.21) (2020-07-19)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.20](https://github.com/dendronhq/dendron/compare/v0.3.19...v0.3.20) (2020-07-18)

### Bug Fixes

- handle special characters in file names when importing using LocalFilePod ([03e42b1](https://github.com/dendronhq/dendron/commit/03e42b167ac9f073cd56f10c4e31b5cecf66dabf))

## [0.3.19](https://github.com/dendronhq/dendron/compare/v0.3.18...v0.3.19) (2020-07-18)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.18](https://github.com/dendronhq/dendron/compare/v0.3.17...v0.3.18) (2020-07-18)

### Features

- ability to import notes from local file system [docs](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.topic.pod.md) ([e44fd11](https://github.com/dendronhq/dendron/commit/e44fd1133bec10f22831f059c8d98cf4076dcdcc))

## [0.3.17](https://github.com/dendronhq/dendron/compare/v0.3.16...v0.3.17) (2020-07-17)

### Features

- more flexible file names for [journal notes](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.special-notes.md) ([9d9f10b](https://github.com/dendronhq/dendron/commit/9d9f10bd0873c201361c4625fc49c9f62ee82991))

## [0.3.16](https://github.com/dendronhq/dendron/compare/v0.3.15...v0.3.16) (2020-07-16)

### Features

- implement journal notes. see details here: https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.feature.journals.md ([5e1236f](https://github.com/dendronhq/dendron/commit/5e1236fddbf1e0fddf4c27d1a40e9841cc99974f))

## [0.3.15](https://github.com/dendronhq/dendron/compare/v0.3.14...v0.3.15) (2020-07-14)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.11](https://github.com/dendronhq/dendron/compare/v0.3.10...v0.3.11) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.10](https://github.com/dendronhq/dendron/compare/v0.1.6...v0.3.10) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.9](https://github.com/dendronhq/dendron/compare/v0.3.8...v0.3.9) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.8](https://github.com/dendronhq/dendron/compare/v0.3.7...v0.3.8) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.7](https://github.com/dendronhq/dendron/compare/v0.3.6...v0.3.7) (2020-07-08)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.6](https://github.com/dendronhq/dendron/compare/v0.3.5...v0.3.6) (2020-07-07)

**Note:** Version bump only for package @dendronhq/common-server

# [0.3.0](https://github.com/dendronhq/dendron/compare/v0.2.20...v0.3.0) (2020-07-05)

**Note:** Version bump only for package @dendronhq/common-server
