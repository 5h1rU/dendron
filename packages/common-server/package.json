{
  "name": "@dendronhq/common-server",
  "version": "0.1.6",
  "description": "common-server",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-server"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com",
    "organization": true
  },
  "keywords": [
    ""
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index",
  "types": "lib/index",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "",
    "lint": "",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean && npm run format && npm run lint && echo Using TypeScript && tsc --version",
    "build": "npm run compile",
    "compile": "tsc -p tsconfig.build.json",
    "test": "jest",
    "coverage": "jest --coverage",
    "watch": "npm run compile -- --watch",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@dendronhq/common-all": "^0.1.6",
    "gray-matter": "^4.0.2",
    "lodash": "^4.17.15",
    "minimatch": "^3.0.4",
    "pino": "^6.3.2",
    "yamljs": "^0.3.0"
  },
  "devDependencies": {
    "@types/pino": "^6.3.0",
    "@types/jest": "^23.3.3",
    "@types/lodash": "^4.14.154",
    "@types/minimatch": "^3.0.3",
    "@types/node": "^10.11.4",
    "@types/yamljs": "^0.2.31",
    "coveralls": "^3.0.2",
    "jest": "^23.6.0",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "ts-jest": "^23.10.3",
    "ts-node": "^7.0.1",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "^3.1.1"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "jest": {
    "preset": "ts-jest"
  },
  "gitHead": "403eb012f8b34bc115dd7dd86f80c76b39329f3b"
}
