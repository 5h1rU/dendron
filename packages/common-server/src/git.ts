import { CONSTANTS, DendronConfig, DVault } from "@dendronhq/common-all";
import fs from "fs-extra";
import path from "path";
import simpleGit, { SimpleGit } from "simple-git";
import { readYAML } from "./files";
import { vault2Path } from "./filesv2";
export { simpleGit, SimpleGit };

// comment
export class GitUtils {
  static getRepoNameFromURL(url: string): string {
    return path.basename(url, ".git");
  }

  static getVaultsFromRepo(opts: {
    repoPath: string;
    wsRoot: string;
  }): { vaults: DVault[] } {
    const { repoPath, wsRoot } = opts;
    // is workspace root
    if (fs.existsSync(path.join(repoPath, CONSTANTS.DENDRON_CONFIG_FILE))) {
      const config = readYAML(
        path.join(repoPath, CONSTANTS.DENDRON_CONFIG_FILE)
      ) as DendronConfig;
      const vaults = config.vaults.map((ent) => {
        const vpath = vault2Path({ vault: ent, wsRoot: repoPath });
        return {
          ...ent,
          fsPath: path.relative(wsRoot, vpath),
        };
      });
      return {
        vaults,
      };
    } else {
      return { vaults: [{ fsPath: path.resolve(wsRoot, repoPath) }] };
    }
  }
}
