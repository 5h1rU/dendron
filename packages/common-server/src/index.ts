export * from "./files";
export * from "./logger";
export { testUtils, FileTestUtils } from "./testUtils";
