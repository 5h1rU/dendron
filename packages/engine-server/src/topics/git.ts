import { DendronError } from "@dendronhq/common-all";
import { findInParent } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import nodegit, { Repository } from "nodegit";
import os from "os";
import path from "path";

const DEFAULT_CREDENTIALS = {
  type: "ssh",
  privateKey: path.join(os.homedir(), ".ssh", "id_rsa"),
  publicKey: path.join(os.homedir(), ".ssh", "id_rsa.pub"),
  passphrase: "",
  username: "",
  password: "",
};

export type Credentials = {
  type: "ssh";
  privateKey: string;
  publicKey: string;
  passphrase: string;
  username: string;
  password: string;
};

export function findRepo() {
  const resp = findInParent(".", ".git");
  if (_.isUndefined(resp)) {
    throw Error("no repo found");
  }
  const pathToRepo = path.join(resp, ".git");
  return nodegit.Repository.open(pathToRepo);
}

export async function isRepo(
  fpath: string
): Promise<nodegit.Repository | false> {
  try {
    return await nodegit.Repository.open(fpath);
  } catch (err) {
    return false;
  }
}

// function onCredentialCheck() {
//   // NOTE: you might instead need to get an SSH key from
//   //    local public/private keys, using `sshKeyNew(..)`.
//   //    either way, the username has to be 'git' if that's
//   //    the username used for the SSH URL for the remote
//   //    (clone, etc)
//   return Cred.sshKeyFromAgent("git");
// }

export class Git {
  static isRepo = isRepo;

  public repo: Repository;

  static createCreds(creds: Partial<Credentials>): Credentials {
    return _.defaults(creds, DEFAULT_CREDENTIALS);
  }

  /**
   * generate options for fetch
   *
   * ```
   * credentials: {
   *   type: string,        // 'ssh' or 'http'
   *   privateKey: string,  // path to ssh private key
   *   publicKey: string,   // path to ssh public key
   *   passphrase: string,  // passphrase of credentials
   *   username: string,    // http username
   *   password: string,    // http password
   * }
   * ```
   *
   * @param {Object} credentials
   * @return {Object}
   */
  static fetchOptions(credentials: Credentials) {
    credentials = Object.assign({}, DEFAULT_CREDENTIALS, credentials);

    const callbacks: any = {};
    // if (process.platform === 'darwin') {
    callbacks.certificateCheck = () => 0;
    // }

    // callbacks.credentials = (_url: string, username: string) => {
    //   if (credentials.type === 'ssh') {
    //     console.log(credentials, username);
    //     return Cred.sshKeyNew(
    //       username,
    //       credentials.publicKey,
    //       credentials.privateKey,
    //       credentials.passphrase
    //     )
    //   } else if (credentials.type === 'http') {
    //     return Cred.userpassPlaintextNew(credentials.username, credentials.password)
    //   } else {
    //     return Cred.usernameNew(username)
    //   }
    // }

    return {
      callbacks,
    };
  }

  static async create(fpath: string) {
    const repo = await isRepo(fpath);
    if (!repo) {
      throw new DendronError({ msg: "no repo found" });
    }
    return new Git(repo);
  }

  static async createRepoWithCommit(fpath: string) {
    const repo = await nodegit.Repository.init(fpath, 0);
    await Git.initRepo(repo);
  }

  static async initRepo(repo: Repository) {
    const root = repo.workdir();
    const gitIgnore = path.join(root, ".gitignore");
    fs.writeFileSync(gitIgnore, "", { encoding: "utf8" });
    const idx = await repo.refreshIndex();
    await idx.addAll();
    await idx.write();
    const oid = await idx.writeTree();
    const author = nodegit.Signature.now("dendron", "bot@dendron.so");
    var committer = nodegit.Signature.now("dendron", "bot@dendron.so");
    await repo.createCommit("HEAD", author, committer, "message", oid, []);
  }

  static async getRemotes(repo: Repository) {
    const refs = await repo.getReferences();
    const remoteRefs = refs.filter((r) => r.isRemote() === 1);
    return remoteRefs.map((r) => r.name());
  }

  static async pushRepo(repo: Repository, credentials: Credentials) {
    // console.log("1", repo, credentials);
    // const branch = await repo.getCurrentBranch()
    // if (!branch.isHead()) {
    //   throw new Error(`HEAD is not pointing to current branch ${branch.shorthand()}, cannot pull`)
    // }
    // console.log("2", branch);

    // const upstream = await Branch.upstream(branch);
    // console.log("3", upstream);
    // const remoteName = await Branch.remoteName(repo, upstream.name())
    // console.log("4");
    // console.log({branch, upstream, remoteName, bond: true});
    // const remote = await this.rawRepo.getRemote(remoteName)
    // const remoteBranchName = upstream.shorthand().replace(`${remoteName}/`, '')
    // const refspec = `refs/heads/${branch.shorthand()}:refs/heads/${remoteBranchName}`

    const remote = await repo.getRemote("refs/remotes/origin/master");
    console.log(remote);
    // NOTE: pulling out all refs so that the push is complete
    const refs = await repo.getReferences();
    const refSpecs = refs.map(function getRefSpec(ref) {
      return `${ref.toString()}:${ref.toString()}`;
    });

    // NOTE: this is how to connect/disconnect the remote, in case
    //    that's needed; but it doesn't seem like it is.
    // await remote.connect(Git.Enums.DIRECTION.PUSH,authenticationCallbacks);
    // await remote.disconnect();

    // NOTE: pushing generally requires authentication, so this is
    //    how to do that:
    await remote.push(refSpecs, {
      ...Git.fetchOptions(credentials),
    });
  }

  constructor(repo: Repository) {
    this.repo = repo;
  }
}

export { nodegit };
