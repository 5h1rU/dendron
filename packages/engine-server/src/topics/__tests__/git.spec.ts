import { FileTestUtils } from "@dendronhq/common-server";
import { nodegit, isRepo } from "../git";
import fs from "fs-extra";
import path from "path";

describe("initRepo", async () => {
  //   test("init repo", async () => {
  //     const root = FileTestUtils.tmpDir().name;
  //     expect(await isRepo(root)).toBeFalsy();
  //   });
});

describe("isRepo", async () => {
  test("no repo", async () => {
    const root = FileTestUtils.tmpDir().name;
    expect(await isRepo(root)).toBeFalsy();
  });

  test("repo", async () => {
    const root = FileTestUtils.tmpDir().name;
    const repo = await nodegit.Repository.init(root, 0);
    const newFilePath = path.join(root, "bond.md");
    fs.writeFileSync(newFilePath, "hello", { encoding: "utf8" });
    const idx = await repo.refreshIndex();
    await idx.addAll();
    await idx.write();
    const oid = await idx.writeTree();
    const author = nodegit.Signature.now("dendron", "bot@dendron.so");
    var committer = nodegit.Signature.now("dendron", "bot@dendron.so");
    const commitId = await repo.createCommit(
      "HEAD",
      author,
      committer,
      "message",
      oid,
      []
    );
    expect(commitId).toMatchSnapshot();
    expect(await isRepo(root)).toBeTruthy();
  });
});
