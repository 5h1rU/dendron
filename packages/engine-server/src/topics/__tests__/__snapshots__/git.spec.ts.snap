// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isRepo repo 1`] = `Oid {}`;
